### PR TITLE
Change Implementation of Validation Loss

### PR DESCRIPTION
**Pull request by linuswalter**
Dear Ehsan,

I changed the workflow on how to treat validation data which leads to an improvement in computation speed during model training. I hope this is a useful contribution to the SciANN Code.
Maybe my implementation has to be rewritten and adapted to the Coding syntax and standards of SciANN.

Best regards,
Linus

---

**Implementation:**
At first, validation data are passed to `model.train` in the same data structure as the training data. Inside the `model.train` function, both datasets are prepared in the same way as well, before being passed to the function `opt_fit_func`.
So basically, I just copied `line 325` until `line 390`and adapted it for the validation loss.

---
**Model Performance:**
The change in performance is quite visible in the computation times per epoch:

1. Model performance without validation loss
Pasted image 20220901155750

2. Model performance with the old validation loss  workflow
Pasted image 20220901162109

3. Model performance with the updated validation loss workflow

### CODE REVIEW

File: sciann/models/model.py [hunk 1]
```diff
@@ -318,6 +319,8 @@ def train(self,
                 . freq: freq of calling CheckPoint callback.
                 . best: If True, only saves the best loss, otherwise save all weights at every `freq` epochs.
             default_zero_weight: a small number for zero sample-weight.
+            weights_val: (List) Weights for the validation loss.
```
Comment (ehsanhaghighat) [inline sciann/models/model.py:322]: I am not sure if why we should use weights for evaluating validation loss - sample weights or adaptive weights are intended to increase training speed -- but validation loss is normally evaluated using uniform weights, right?

File: sciann/models/model.py [hunk 2]
```diff
@@ -388,6 +391,83 @@ def train(self,
                 x_true, y_star, sample_weights, 
                 batch_size, shuffle
             )
+# =============================================================================
+# Prepare Validation Data similar to Training Data
```
Comment (ehsanhaghighat) [inline sciann/models/model.py:395]: agree that it is better to be consistent - make sense - thanks Walter

File: sciann/models/model.py (hunk 1 above)
Comment (linuswalter) [reply sciann/models/model.py:322]: Actually, I just introduced the validation weights because I wasn't sure if one should assign some. And now, when thinking about it, it seems more consistent to assign the same weights for training and validation? Because, weighting has an influence on the total loss, right? And I think training and validation loss should be calculated  in the same way in order to be able to compare the loss curves of training and validation.
Consequently, we should the option of assigning extra weights for the validation loss (i.e. my implementation was wrong), but use instead the training weights internally also as validation weights.
@ehsanhaghighat , what do you think?
But I am also not sure about it, I can do a little research about it, if you want. :thinking: 
Best, Linus
